### DISCUSSION TITLE
resource with lifetime

### DISCUSSION THREAD

**Discussion by dvic** [rusterlium/rustler#535] [Q&A]
Is it possible to have a resource that hands out another resource bound by the lifetime of the first resource? It seems I can't work with lifetimes at all when defining resources.

In my specific usecase I'm trying to return an iterator (a resource) from a RocksDB instance (a resource), and the lifetime of the iterator is tied to the lifetime of the RocksDB instance.

**Comment by filmor** [comment]
Could you show a bit of code to clarify what you would like to do? Lifetimes of `ResourceArc` objects is tied to the lifetime of the given Erlang environment, I'm not sure how one could even express what (I think) you want in Rust.

**Reply by dvic** [reply]
Sure: I'm trying to get the `rocker` library up to date. Here are the two resources:  https://github.com/Vonmo/rocker/compare/master...dvic:rocker:dvic-bump-deps#diff-f7a3db49ff9c7d28c32196b4004f0b12a096c155a3b68e157d627b19cc6bbe76R25-R32

Obviously that can't be 'static but that's the issue i'm running into. The iterator function (https://github.com/Vonmo/rocker/compare/master...dvic:rocker:dvic-bump-deps#diff-f7a3db49ff9c7d28c32196b4004f0b12a096c155a3b68e157d627b19cc6bbe76R311) is where the iterator resource is created and then it's queried using the next function (https://github.com/Vonmo/rocker/compare/master...dvic:rocker:dvic-bump-deps#diff-f7a3db49ff9c7d28c32196b4004f0b12a096c155a3b68e157d627b19cc6bbe76R372). 

What is the recommended pattern here? When you want to expose an iterator interface and the library in question has a lifetime on the iterator :)

**Reply by dvic** [reply]
And the reason it's tied to the DB instance is because iterator eventually contains a `DBRawIteratorWithThreadMode` and that one uses phantom data to signal the lifetime link: https://docs.rs/rocksdb/latest/src/rocksdb/db_iterator.rs.html#73-87

**Reply by filmor** [reply]
Not sure whether that is a good idea, but you can theoretically have `Iterator` include a `ResourceArc` member which will ensure that the `DB` is definitely alive while the `Iterator` is alive. That way it should be safe to use `unsafe` blocks to create the iterator reference.

**Reply by dvic** [reply]
But would the `IteratorResource` then have a lifetime on the struct definition?
